docs/design/2018-08-29-new-planner.md

```diff
@@ -0,0 +1,411 @@
+each equivalent expression is:
+
+```go
+// handle implementation rules for each equivalent expression.
+for _, curExpr := range g.equivalents {
+	for _, impl := range implGroupExpr(curExpr) {
+		impl.setCumCost(impl.getSelfCost())
+		for i, childGroup := range curExpr.children {
+			childImpl := implGroup(childGroup, impl.getChildReqProp(i), costLimit-impl.getCumCost())
+			if childImpl == nil {
+				cumCost = math.MaxFloat64
+				break
+			}
+			impl.setChild(i, childImpl)
+			impl.setCumCost(impl.getCumCost() + childImpl.getCumCost())
+		}
+		if groupImpl.getCumCost() > impl.getCumCost() {
+			groupImpl = impl
+		}
+	}
+}
+```
+
+To enumerate all the implementation for an expression, we have to enumerate all
+the applicable implementation rules on that expression, and calculate the
+self-cost for physical implementation:
+
+```go
+func implGroupExpr(cur *GroupExpr, reqPhysProp *PhysicalProperty) (impls []PhysicalPlan) {
+	for _, rule := range GetImplementationRules(cur.exprNode) {
+		if !rule.match(cur, reqPhysProp) {
+			continue
+		}
+		impl := rule.onImplement(cur)
+		impl.calcSelfCost(cur.children)
+		impls = append(impls, impl)
+	}
+	return impls
+}
+```
+
+Also, the enforcing rules should be considered:
+
+```go
+// handle enforcing rules for the required physical property.
+for _, rule := range getEnforcerRules(reqPhysProp) {
+	impl, newReqPhysProp := rule.onEnforce(reqPhysProp)
+	impl.calcSelfCost(g)
+	impl.setCumCost(impl.getSelfCost())
+	childImpl := implGroup(g, newReqPhysProp, costLimit-impl.getCumCost())
+	impl.setChild(0, childImpl)
+	impl.setCumCost(impl.getCumCost() + childImpl.getCumCost())
+	if groupImpl.getCumCost() > impl.getCumCost() {
+		groupImpl = impl
+	}
+}
+```
+
+## Rationale
+
+## Compatibility
+
+No
+
+## Implementation
+
+The first step to implement the new planner is to add a session variable named
+`tidb_enable_volcano_planner` to control whether to use the new planner. Once
+this variable is set, all the optimization steps are handed to the new planner.
+The procedure of converting the abstract syntax tree to a logical plan is
+remained unchanged. The constructed physical plan is also compatible to the
+existing physical plan. The only affection should be the optimization algorithm.
+
+The second step is implementing the framework of the new planner, including the
+conceptions described above: `Group`/`GroupExpr`/`ExprIter`/`Pattern`, the
+interfaces like `transformation`/`implementation`/`enforcing`, and the functions
+like `exploreGroup`/`findMoreEquiv`/`implGroup`/`implGroupExpr`.
+
+The third step is to add some simple transformation/implementation/enforcing
+rules and tests to make the new planner framework basically available.
+
+The fourth step is to adopt the "Adaptor" conception to rewrite the operator
+push-down logical for different storages.
+
+The fifth step is to add some rules which are not able or not easy to be added
```

Comment:
The fifth step is to add some rules which are not **allowed** or not easy to be added?
or
The fifth step is to add some rules which **are not easy or cannot** to be added?